### ISSUE TITLE
Reference to local variable creates unwanted dependencies to resources in another instance of a module.

### ISSUE THREAD

**Issue by suvaanshkumar** [hashicorp/terraform#32108]
### Terraform Version

```shell
Terraform v1.3.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/random v3.4.3
+ provider registry.terraform.io/hashicorp/time v0.9.0
```

### Terraform Configuration Files

main.tf
```terraform
locals {
  items = {
    "fourth" = {
      "wait" = 13
    },
    "first" = {
      "wait" = 1
    },
    "second" = {
      "wait" = 2
    },
    "third" = {
      "wait" = 5
    }
  }
}


module "mymodule" {
  for_each = local.items
  source = "./mymodule"
  wait  = each.value.wait
}
```

mymodule/main.tf:
```terraform
variable "wait" {
  type        = number
  default     = 0
}

locals {
    cluster_id = time_sleep.part1.id
}

resource "time_sleep" "part1" {
    create_duration = "${var.wait}s"
}

resource "random_pet" "part2" {
    # This produces incorrect behaviour (all part1 must be finished before any part2 starts creating):
    prefix = "PREFIX:${local.cluster_id}"

    # This produces the correct behaviour:
    #prefix = "PREFIX:${time_sleep.part1.id}"
}
```

### Debug Output

https://gist.github.com/suvaanshkumar/ae7a918c951dd5ec0c8d8ca071439b02

### Expected Behavior

The part2 resource in mymodule should start to be created as soon as the part1 resources is created.

For example:

```
module.mymodule["first"].time_sleep.part1: Creating...
module.mymodule["second"].time_sleep.part1: Creating...
module.mymodule["fourth"].time_sleep.part1: Creating...
module.mymodule["third"].time_sleep.part1: Creating...
module.mymodule["first"].time_sleep.part1: Creation complete after 1s [id=2022-10-27T22:20:34Z]
module.mymodule["first"].random_pet.part2: Creating...
module.mymodule["first"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:20:34Z-genuine-sheep]
module.mymodule["second"].time_sleep.part1: Creation complete after 2s [id=2022-10-27T22:20:35Z]
module.mymodule["second"].random_pet.part2: Creating...
module.mymodule["second"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:20:35Z-novel-mole]
module.mymodule["third"].time_sleep.part1: Creation complete after 5s [id=2022-10-27T22:20:38Z]
module.mymodule["third"].random_pet.part2: Creating...
module.mymodule["third"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:20:38Z-alive-bug]
module.mymodule["fourth"].time_sleep.part1: Still creating... [10s elapsed]
module.mymodule["fourth"].time_sleep.part1: Creation complete after 13s [id=2022-10-27T22:20:46Z]
module.mymodule["fourth"].random_pet.part2: Creating...
module.mymodule["fourth"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:20:46Z-moral-parrot]
```

The `random_pet.part2` resource for the "first" instance of the module is created as soon as the `time_sleep.part1` for the "first" instance of the module is completed, and so on for "second", "third", and "fourth" instances of the module.

### Actual Behavior

The part2 resources for all modules wait until all part1 resources have finished creating.

For example:

```
module.mymodule["second"].time_sleep.part1: Creating...
module.mymodule["third"].time_sleep.part1: Creating...
module.mymodule["first"].time_sleep.part1: Creating...
module.mymodule["fourth"].time_sleep.part1: Creating...
module.mymodule["first"].time_sleep.part1: Creation complete after 1s [id=2022-10-27T22:19:45Z]
module.mymodule["second"].time_sleep.part1: Creation complete after 2s [id=2022-10-27T22:19:46Z]
module.mymodule["third"].time_sleep.part1: Creation complete after 5s [id=2022-10-27T22:19:49Z]
module.mymodule["fourth"].time_sleep.part1: Still creating... [10s elapsed]
module.mymodule["fourth"].time_sleep.part1: Creation complete after 13s [id=2022-10-27T22:19:57Z]
module.mymodule["first"].random_pet.part2: Creating...
module.mymodule["third"].random_pet.part2: Creating...
module.mymodule["second"].random_pet.part2: Creating...
module.mymodule["fourth"].random_pet.part2: Creating...
module.mymodule["second"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:19:46Z-normal-silkworm]
module.mymodule["third"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:19:49Z-easy-rooster]
module.mymodule["first"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:19:45Z-internal-caribou]
module.mymodule["fourth"].random_pet.part2: Creation complete after 0s [id=PREFIX:2022-10-27T22:19:57Z-glad-squirrel]
```

Above, all of the `random_pet.part2` resources are delayed until `time_sleep.part1` for the "fourth" instance has completed.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

To fix the error, do not reference the local, e.g. edit mymodule/main.tf and:
1. Comment the line `prefix = "PREFIX:${local.cluster_id}"`
2. Uncomment the line `#prefix = "PREFIX:${time_sleep.part1.id}"`
3. `terraform destroy`
4. `terraform apply`

**Comment by apparentlymart** [comment]
Hi @suvaanshkumar! Thanks for reporting this.

What you have described sounds like the intended behavior of Terraform. Dependencies are always between entire static declarations and not between individual instances of those declarations, because the `for_each` expression `local.items` itself has dependencies (on the "items" local value) and so Terraform must build the dependency graph long before it knows which instances of that module exist.

Although the observed behavior is a bit different in your case, because the resulting dependency graph does not contain any cycles, it seems like this is the same root concern as we were discussing in #31072 and so I'm going to close this one just to consolidate the discussion over there. That issue is currently representing the possibility of Terraform using some other strategy for analyzing dependencies that might allow it to create a more detailed dynamic graph for individual instances, but there is not yet any known design to achieve that and so that issue is on hold until there is a known design and until there are the resources to implement that design.

Thanks!
